# Prediction of Wine Quality

### Parameters present in wine-quality csv files:
#### 1. Fixed Acidity <br> 2. Volatile Acidity <br> 3. Citric Acid <br> 4. Residual Sugar <br> 5. Chlorides <br> 6. Free Sulfur Dioxide <br> 7. Total Sulfur Dioxide <br> 8. Density <br> 9. pH <br> 10. Sulphates <br> 11. Alcohol Percentage (%) <br> 12. Quality

## Preliminary Data Exploration

In [1]:
import pandas as pd
import numpy as np

In [2]:
# We start with White Wine
df = pd.read_csv("winequality-white.csv")

In [4]:
df.info()
df.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4898 entries, 0 to 4897
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         4898 non-null   float64
 1   volatile acidity      4898 non-null   float64
 2   citric acid           4898 non-null   float64
 3   residual sugar        4898 non-null   float64
 4   chlorides             4898 non-null   float64
 5   free sulfur dioxide   4898 non-null   float64
 6   total sulfur dioxide  4898 non-null   float64
 7   density               4898 non-null   float64
 8   pH                    4898 non-null   float64
 9   sulphates             4898 non-null   float64
 10  alcohol               4898 non-null   float64
 11  quality               4898 non-null   int64  
dtypes: float64(11), int64(1)
memory usage: 459.3 KB


fixed acidity  volatile acidity  citric acid  residual sugar  \
count    4898.000000       4898.000000  4898.000000     4898.000000   
mean        6.854788          0.278241     0.334192        6.391415   
std         0.843868          0.100795     0.121020        5.072058   
min         3.800000          0.080000     0.000000        0.600000   
25%         6.300000          0.210000     0.270000        1.700000   
50%         6.800000          0.260000     0.320000        5.200000   
75%         7.300000          0.320000     0.390000        9.900000   
max        14.200000          1.100000     1.660000       65.800000   

         chlorides  free sulfur dioxide  total sulfur dioxide      density  \
count  4898.000000          4898.000000           4898.000000  4898.000000   
mean      0.045772            35.308085            138.360657     0.994027   
std       0.021848            17.007137             42.498065     0.002991   
min       0.009000             2.000000              9.000000     0.987110   
25%       0.036000            23.000000            108.000000     0.991723   
50%       0.043000            34.000000            134.000000     0.993740   
75%       0.050000            46.000000            167.000000     0.996100   
max       0.346000           289.000000            440.000000     1.038980   

                pH    sulphates      alcohol      quality  
count  4898.000000  4898.000000  4898.000000  4898.000000  
mean      3.188267     0.489847    10.514267     5.877909  
std       0.151001     0.114126     1.230621     0.885639  
min       2.720000     0.220000     8.000000     3.000000  
25%       3.090000     0.410000     9.500000     5.000000  
50%       3.180000     0.470000    10.400000     6.000000  
75%       3.280000     0.550000    11.400000     6.000000  
max       3.820000     1.080000    14.200000     9.000000

### Summary:<br>We have 12 columns including the quality. From the looks of it, all the parameters are continuous. Linear Regression could be used as a good preliminary test for setting a baseline. 

## Linear Regression - Prediction of Quality using the Parameters

In [6]:
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error, r2_score

In [14]:
df_X = df.copy()
df_X = df_X.drop(["quality"],axis=1)
df_y = df['quality']

X_train, X_test, y_train, y_test = train_test_split(df_X, df_y, test_size=0.2, random_state=1)

linear_regression1 = LinearRegression()
linear_regression1.fit(X_train, y_train)

y_pred = linear_regression1.predict(X_test)

mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print("Mean Squared Error of:",mse,"\nR2 Square:",r2)

def mean_absolute_percentage_error(y_true, y_pred): 
    y_true, y_pred = np.array(y_true), np.array(y_pred)
    return np.mean(np.abs((y_true - y_pred) / y_true)) * 100

mape = mean_absolute_percentage_error(y_test, y_pred)
print(f'Mean Absolute Percentage Error (MAPE): {mape:.2f}%')
print(f'Accuracy based on MAPE: {100-mape:.2f}%')

Mean Squared Error of: 0.542548726356103 
R2 Square: 0.2901221946674388
Mean Absolute Percentage Error (MAPE): 10.20%
Accuracy based on MAPE: 89.80%


### Summary:<br> Based on LinearRegression we have received a prediction accuracy of 89.8%.

## Linear Regression - Feature Importance

In [17]:
linear_regression2 = LinearRegression()
linear_regression2.fit(df_X, df_y)
feature_importance = pd.DataFrame({'Feature': df_X.columns, 'Coefficient': linear_regression2.coef_})
feature_importance = feature_importance.sort_values(by='Coefficient', ascending=False)
print(feature_importance)

                 Feature  Coefficient
8                     pH     0.686344
9              sulphates     0.631476
10               alcohol     0.193476
3         residual sugar     0.081483
0          fixed acidity     0.065520
2            citric acid     0.022090
5    free sulfur dioxide     0.003733
6   total sulfur dioxide    -0.000286
4              chlorides    -0.247277
1       volatile acidity    -1.863177
7                density  -150.284181
